tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/fixer/TableNotFoundFixer.java

```diff
@@ -0,0 +1,122 @@
+package com.google.cloud.bigquery.utils.queryfixer.fixer;
+
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.utils.queryfixer.QueryPositionConverter;
+import com.google.cloud.bigquery.utils.queryfixer.entity.FixOption;
+import com.google.cloud.bigquery.utils.queryfixer.entity.FixResult;
+import com.google.cloud.bigquery.utils.queryfixer.entity.Position;
+import com.google.cloud.bigquery.utils.queryfixer.errors.TableNotFoundError;
+import com.google.cloud.bigquery.utils.queryfixer.service.BigQueryService;
+import com.google.cloud.bigquery.utils.queryfixer.util.PatternMatcher;
+import com.google.cloud.bigquery.utils.queryfixer.util.StringUtil;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * The fixer class responsible for "table not found" error. It fixes the error by (1) find the
+ * similar table under the same project and dataset as the incorrect one, and (2) replace the
+ * incorrect one with the most similar table.
+ *
+ * <p>If no similar tables are found in the dataset, the fixer will directly return the error
+ * without providing any fix options.
+ */
+public class TableNotFoundFixer implements IFixer {
+
+  /** This regex is able to parse both project:dataset.table and dataset.table. */
+  private static final String tableIdRegex = "^((.*?):)?(.*?)\\.(.*?)$";
+
+  private final String query;
+  private final TableNotFoundError err;
+  private final BigQueryService bigQueryService;
+  private final QueryPositionConverter queryPositionConverter;
+
+  public TableNotFoundFixer(String query, TableNotFoundError err, BigQueryService bigQueryService) {
+    this.query = query;
+    this.err = err;
+    this.bigQueryService = bigQueryService;
+    this.queryPositionConverter = new QueryPositionConverter(query);
+  }
+
+  @Override
+  public FixResult fix() {
+    TableId fullTableId = constructTableId(err.getTableName());
+    List<String> tableNames =
+        bigQueryService.listTableNames(fullTableId.getProject(), fullTableId.getDataset());
+
+    StringUtil.SimilarStrings similarTables =
+        StringUtil.findSimilarWords(tableNames, fullTableId.getTable());
+
+    // this is an arbitrary standard. It requires the candidate table should share at least 50%
+    // similarity as
+    // the incorrect table typo.
+    int editDistanceThreshold = (fullTableId.getTable().length() + 1) / 2;
```

Comment:
maybe add a TODO to make this user configurable.

Reply:
Done